File: java/var-core/src/main/java/dev/varar/core/Hash.java
package dev.varar.core;

/**
 * FNV-1a (32-bit) change-detector over UTF-16 code units. Port of {@code
 * var-core/src/hash.ts}; byte-identical to the TypeScript and Python ports so
 * {@code varar.lock.json} fingerprints match across every language. Java {@code
 * char} is already a UTF-16 code unit (like JS {@code charCodeAt}), and {@code
 * int} arithmetic wraps mod 2^32, so this is a direct transliteration. The
 * {@code fnv1a:} prefix namespaces the algorithm.
 */
public final class Hash {

    private Hash() {}

    private static final int FNV_OFFSET = 0x811c9dc5;
    private static final int FNV_PRIME = 0x01000193;

    /** Mirrors {@code hashSource()} from hash.ts. */
    public static String hashSource(String source) {
        int h = FNV_OFFSET;
        for (int i = 0; i < source.length(); i++) {
            h = (h ^ source.charAt(i)) * FNV_PRIME;
        }
        // %08x formats the 32-bit pattern as unsigned lowercase hex.
        return String.format("fnv1a:%08x", h);
    }
}

File: dotnet/Varar.Core/Hash.cs
namespace Varar.Core;

public static class Hash
{
    /// <summary>
    /// FNV-1a (32-bit) change-detector over UTF-16 code units. Not a security hash — tiny and
    /// trivially portable. The <c>fnv1a:</c> prefix namespaces the algorithm. Port of <c>hash.ts</c>.
    /// </summary>
    public static string HashSource(string source)
    {
        uint h = 0x811c9dc5;
        foreach (char c in source)
        {
            h ^= c;
            h *= 0x01000193u; // uint multiply wraps mod 2^32 (unchecked by default)
        }

        return $"fnv1a:{h:x8}";
    }
}
